File: src/lib/storage.js
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';

const DATA_DIR = path.join(process.cwd(), 'data');

const getFilePath = (fileName) => path.join(DATA_DIR, fileName);

const readData = (fileName) => {
  try {
    const filePath = getFilePath(fileName);
    if (!fs.existsSync(filePath)) {
      return [];
    }
    const data = fs.readFileSync(filePath, 'utf8');
    return JSON.parse(data);
  } catch (error) {
    return [];
  }
};

const writeData = (fileName, data) => {
  const filePath = getFilePath(fileName);
  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
  }
  fs.writeFileSync(filePath, JSON.stringify(data, null, 2), 'utf8');
};

const calculateTier = (points) => {
  if (points < 50) return 'Novice';
  if (points < 150) return 'Expert';
  if (points < 300) return 'Master';
  return 'Grandmaster';
};

export const getQuestions = (subject) => {
  const questions = readData('questions.json');
  let filtered = questions;
  if (subject) {
    filtered = questions.filter(q => q.subject === subject);
  }
  return filtered.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
};

export const getQuestionById = (id) => {
  const questions = readData('questions.json');
  return questions.find(q => q.id === id) || null;
};

export const createQuestion = (data) => {
  const questions = readData('questions.json');
  const newQuestion = {
    ...data,
    id: crypto.randomUUID(),
    createdAt: new Date().toISOString(),
    aiAnswer: data.mode === 'ai' ? data.aiAnswer || null : null,
    status: 'open'
  };
  questions.push(newQuestion);
  writeData('questions.json', questions);
  return newQuestion;
};

export const getAnswersByQuestionId = (questionId) => {
  const answers = readData('answers.json');
  const questionAnswers = answers.filter(a => a.questionId === questionId);
  return questionAnswers.sort((a, b) => {
    if (a.isBestFit && !b.isBestFit) return -1;
    if (!a.isBestFit && b.isBestFit) return 1;
    return b.votes - a.votes;
  });
};

export const createAnswer = (data) => {
  const answers = readData('answers.json');
  const newAnswer = {
    ...data,
    id: crypto.randomUUID(),
    votes: 0,
    isBestFit: false,
    bestFitReason: null,
    createdAt: new Date().toISOString()
  };
  answers.push(newAnswer);
  writeData('answers.json', answers);
  return newAnswer;
};

export const updateAnswer = (id, updates) => {
  const answers = readData('answers.json');
  const index = answers.findIndex(a => a.id === id);
  if (index === -1) return null;
  answers[index] = { ...answers[index], ...updates };
  writeData('answers.json', answers);
  return answers[index];
};

export const getUsers = () => {
  const users = readData('users.json');
  return users.sort((a, b) => b.points - a.points);
};

export const getOrCreateUser = (name, role) => {
  const users = readData('users.json');
  let user = users.find(u => u.name === name);
  if (!user) {
    user = {
      id: crypto.randomUUID(),
      name,
      role: role || 'student',
      points: 0,
      tier: 'Novice'
    };
    users.push(user);
    writeData('users.json', users);
  }
  return user;
};

export const updateUserPoints = (name, pointsDelta) => {
  const users = readData('users.json');
  const index = users.findIndex(u => u.name === name);
  if (index === -1) return null;
  users[index].points += pointsDelta;
  users[index].tier = calculateTier(users[index].points);
  writeData('users.json', users);
  return users[index];
};

export const voteAnswer = (answerId, direction) => {
  const answers = readData('answers.json');
  const index = answers.findIndex(a => a.id === answerId);
  if (index === -1) return null;
  
  const answer = answers[index];
  if (direction === 'up') {
    answer.votes += 1;
  } else if (direction === 'down') {
    answer.votes -= 1;
  }
  
  writeData('answers.json', answers);
  
  const authorName = answer.authorName;
  const authorRole = answer.authorRole;
  getOrCreateUser(authorName, authorRole);
  const pointsDelta = direction === 'up' ? 10 : direction === 'down' ? -5 : 0;
  if (pointsDelta !== 0) {
    updateUserPoints(authorName, pointsDelta);
  }
  
  return answer;
};

export const markBestFit = (questionId, answerId, reason) => {
  const answers = readData('answers.json');
  const questions = readData('questions.json');
  
  let targetAnswer = null;
  answers.forEach(a => {
    if (a.questionId === questionId) {
      if (a.id === answerId) {
        a.isBestFit = true;
        a.bestFitReason = reason;
        targetAnswer = a;
      } else {
        a.isBestFit = false;
        a.bestFitReason = null;
      }
    }
  });
  
  if (targetAnswer) {
    writeData('answers.json', answers);
    
    const qIndex = questions.findIndex(q => q.id === questionId);
    if (qIndex !== -1) {
      questions[qIndex].status = 'resolved';
      writeData('questions.json', questions);
    }
    
    getOrCreateUser(targetAnswer.authorName, targetAnswer.authorRole);
    updateUserPoints(targetAnswer.authorName, 25);
  }
  
  return targetAnswer;
}

File: app/services/storage.py
import json
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import List, Optional, Dict, Any
from app.config import DATA_DIR
from app.models import Question, Answer, User, TierType

QUESTIONS_FILE = DATA_DIR / "questions.json"
ANSWERS_FILE = DATA_DIR / "answers.json"
USERS_FILE = DATA_DIR / "users.json"

def _read_json(file_path: Path) -> list:
    if not file_path.exists():
        return []
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception as e:
        print(f"Error reading {file_path}: {e}")
        return []

def _write_json(file_path: Path, data: list):
    try:
        with open(file_path, "w", encoding="utf-8") as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
    except Exception as e:
        print(f"Error writing to {file_path}: {e}")

def calculate_tier(points: int) -> TierType:
    if points < 50:
        return "Novice"
    elif points < 150:
        return "Expert"
    elif points < 300:
        return "Master"
    return "Grandmaster"

# ----------------- QUESTIONS -----------------
def get_questions(subject: Optional[str] = None) -> List[Dict[str, Any]]:
    questions = _read_json(QUESTIONS_FILE)
    if subject and subject != "All":
        questions = [q for q in questions if q.get("subject") == subject]
    # Sort by created_at desc
    return sorted(questions, key=lambda q: q.get("created_at", ""), reverse=True)

def get_question_by_id(question_id: str) -> Optional[Dict[str, Any]]:
    questions = _read_json(QUESTIONS_FILE)
    for q in questions:
        if q.get("id") == question_id:
            return q
    return None

def create_question(data: Dict[str, Any]) -> Dict[str, Any]:
    questions = _read_json(QUESTIONS_FILE)
    new_id = str(uuid.uuid4())
    now_iso = datetime.now(timezone.utc).isoformat()
    
    question = {
        "id": new_id,
        "title": data["title"],
        "body": data["body"],
        "subject": data["subject"],
        "author_name": data.get("author_name", "Anonymous Student"),
        "author_role": data.get("author_role", "student"),
        "created_at": now_iso,
        "ai_answer": data.get("ai_answer", None),
        "mode": data.get("mode", "mentor"),
        "status": "open"
    }
    questions.append(question)
    _write_json(QUESTIONS_FILE, questions)
    return question

def update_question_ai_answer(question_id: str, ai_answer: str):
    questions = _read_json(QUESTIONS_FILE)
    for q in questions:
        if q.get("id") == question_id:
            q["ai_answer"] = ai_answer
            break
    _write_json(QUESTIONS_FILE, questions)

# ----------------- ANSWERS -----------------
def get_answers_by_question_id(question_id: str) -> List[Dict[str, Any]]:
    answers = _read_json(ANSWERS_FILE)
    q_answers = [a for a in answers if a.get("question_id") == question_id or a.get("questionId") == question_id]
    
    # Normalize keys
    for a in q_answers:
        if "questionId" in a and "question_id" not in a:
            a["question_id"] = a["questionId"]
        if "authorName" in a and "author_name" not in a:
            a["author_name"] = a["authorName"]
        if "authorRole" in a and "author_role" not in a:
            a["author_role"] = a["authorRole"]
        if "isBestFit" in a and "is_best_fit" not in a:
            a["is_best_fit"] = a["isBestFit"]
        if "bestFitReason" in a and "best_fit_reason" not in a:
            a["best_fit_reason"] = a["bestFitReason"]
        if "createdAt" in a and "created_at" not in a:
            a["created_at"] = a["createdAt"]

    # Sort: best-fit first, then by votes desc
    return sorted(q_answers, key=lambda a: (1 if a.get("is_best_fit") else 0, a.get("votes", 0)), reverse=True)

def create_answer(data: Dict[str, Any]) -> Dict[str, Any]:
    answers = _read_json(ANSWERS_FILE)
    new_id = str(uuid.uuid4())
    now_iso = datetime.now(timezone.utc).isoformat()
    
    answer = {
        "id": new_id,
        "question_id": data["question_id"],
        "body": data["body"],
        "author_name": data.get("author_name", "Anonymous Mentor"),
        "author_role": data.get("author_role", "senior"),
        "votes": 0,
        "is_best_fit": False,
        "best_fit_reason": None,
        "created_at": now_iso
    }
    answers.append(answer)
    _write_json(ANSWERS_FILE, answers)
    
    # Ensure user exists in leaderboard
    get_or_create_user(answer["author_name"], answer["author_role"])
    
    return answer

def vote_answer(answer_id: str, direction: str) -> Optional[Dict[str, Any]]:
    answers = _read_json(ANSWERS_FILE)
    target = None
    for a in answers:
        if a.get("id") == answer_id:
            target = a
            break
            
    if not target:
        return None
        
    delta = 1 if direction == "up" else -1
    target["votes"] = target.get("votes", 0) + delta
    _write_json(ANSWERS_FILE, answers)
    
    # Update author's Kaggle karma (+10 for upvote, -5 for downvote)
    author_name = target.get("author_name") or target.get("authorName")
    author_role = target.get("author_role") or target.get("authorRole") or "senior"
    if author_name:
        points_delta = 10 if direction == "up" else -5
        update_user_points(author_name, points_delta, author_role)
        
    return target

def mark_best_fit(question_id: str, answer_id: str, reason: str) -> Optional[Dict[str, Any]]:
    answers = _read_json(ANSWERS_FILE)
    questions = _read_json(QUESTIONS_FILE)
    
    target_answer = None
    for a in answers:
        q_id = a.get("question_id") or a.get("questionId")
        if q_id == question_id:
            if a.get("id") == answer_id:
                a["is_best_fit"] = True
                a["best_fit_reason"] = reason
                target_answer = a
            else:
                a["is_best_fit"] = False
                a["best_fit_reason"] = None
                
    if target_answer:
        _write_json(ANSWERS_FILE, answers)
        
        # Mark question resolved
        for q in questions:
            if q.get("id") == question_id:
                q["status"] = "resolved"
                break
        _write_json(QUESTIONS_FILE, questions)
        
        # Award author +25 bonus points
        author_name = target_answer.get("author_name") or target_answer.get("authorName")
        author_role = target_answer.get("author_role") or target_answer.get("authorRole") or "senior"
        if author_name:
            update_user_points(author_name, 25, author_role)
            
    return target_answer

# ----------------- USERS & LEADERBOARD -----------------
def get_users() -> List[Dict[str, Any]]:
    users = _read_json(USERS_FILE)
    return sorted(users, key=lambda u: u.get("points", 0), reverse=True)

def get_or_create_user(name: str, role: str = "student") -> Dict[str, Any]:
    users = _read_json(USERS_FILE)
    for u in users:
        if u.get("name") == name:
            return u
            
    new_user = {
        "id": str(uuid.uuid4()),
        "name": name,
        "role": role,
        "points": 0,
        "tier": "Novice"
    }
    users.append(new_user)
    _write_json(USERS_FILE, users)
    return new_user

def update_user_points(name: str, points_delta: int, role: str = "senior") -> Dict[str, Any]:
    users = _read_json(USERS_FILE)
    user = None
    for u in users:
        if u.get("name") == name:
            user = u
            break
            
    if not user:
        user = {
            "id": str(uuid.uuid4()),
            "name": name,
            "role": role,
            "points": 0,
            "tier": "Novice"
        }
        users.append(user)
        
    user["points"] = max(0, user.get("points", 0) + points_delta)
    user["tier"] = calculate_tier(user["points"])
    _write_json(USERS_FILE, users)
    return user
